# 1. What is the Big-O notation for the worst-case time complexity of the following function?

In [1]:

def find_element(arr, x):
    for element in arr:
        if element == x:
            return True
    return False
animal_list = ["Lion", "Elephant", "Tiger", "Giraffe", "Zebra", "Panda", "Kangaroo", "Penguin"]

find_element(animal_list , "Penguin")

True

# 2.  Rearrange the following lines to correctly implement a bubble sort algorithm for a list.

# What is the output if…

In [ ]:


def bubble_sort(items):
return items
for j in range(len(items)-1-i):
if items[j] > items[j+1]:
""" Implementation of bubble sort."""
items[j], items[j+1] = items[j+1], items[j]
for i in range(len(items)):

Output = [64, 34, 25, 12, 22, 11, 90] for [64, 34, 25, 12, 22, 11, 90]

Output = [11, 12, 22, 25, 34, 64, 90] for (64, 34, 25, 12, 22, 11, 90)
Output = [11, 12, 22, 25, 34, 64, 90] for [64, 34, 25, 12, 22, 11, 90
Output = [11, 12, 22, 25, 36, 64, 90] for 64, 34, 25, 12, 22, 11, 90

# Answer
Among the options provided:

[64, 34, 25, 12, 22, 11, 90] - This is the unsorted input, not the output.
[11, 12, 22, 25, 34, 64, 90] - This is the correct output for the given input.
[11, 12, 22, 25, 34, 64, 90] - This is also correct, but the input format is slightly different (missing a closing bracket).
[11, 12, 22, 25, 36, 64, 90] - This is incorrect as it includes 36 instead of 34.

Therefore, the correct answer is:
Output = [11, 12, 22, 25, 34, 64, 90] for [64, 34, 25, 12, 22, 11, 90]

# Rearranged code 

In [2]:
def bubble_sort(items):
    """ Implementation of bubble sort."""
    
    for i in range(len(items)):
        for j in range(len(items)-1-i):
            if items[j] > items[j+1]:
                items[j], items[j+1] = items[j+1], items[j]
    return items

3. Rearrange the following code to correctly define a Python class representing a BankAccount with a method to deposit funds.

Which of the following lines of code will open a new bank account with a balance of 100, deposit an additional 50, and print the balance?

In [ ]:

def deposit(self, amount):
 self.balance += amount
class BankAccount:
 def __init__(self, balance):
    self.balance = balance


# Output objectives 

In [ ]:
150 for account = BankAccount(100); account.deposit(50); print(account.balance)

200 for account.deposit(50); BankAccount(100); print(account.balance)

150 for account.deposit(50); account = BankAccount(150); print(account.balance)

TypeError for account = BankAccount(150); account.deposit('50'); print(account.balance)) 

# Answer 

In [ ]:
class BankAccount:
    def __init__(self, balance):
        self.balance = balance
    
    def deposit(self, amount):
        self.balance += amount

In [ ]:
 #the correct answer is:
150 for account = BankAccount(100); account.deposit(50); print(account.balance)


# Explanation

Now, let's analyze the given options to determine which one correctly opens a new bank account with a balance of 100, deposits an additional 50, and prints the balance:

150 for account = BankAccount(100); account.deposit(50); print(account.balance)
This is correct. It creates an account with 100, deposits 50, and prints 150.
200 for account.deposit(50); BankAccount(100); print(account.balance)
This is incorrect. It tries to use account before it's defined, and the order is wrong.
150 for account.deposit(50); account = BankAccount(150); print(account.balance)
This is incorrect. It tries to deposit before the account is created, and the initial balance is wrong.
TypeError for account = BankAccount(150); account.deposit('50'); print(account.balance))
This would indeed raise a TypeError because '50' is a string, not a number.

Therefore, the correct answer is:
150 for account = BankAccount(100); account.deposit(50); print(account.balance)
This line of code will:

Create a new BankAccount with an initial balance of 100
Deposit an additional 50 to the account
Print the balance, which will be 150

Rearrange the following lines to create a function that uses recursion to reverse a string.

Which of the following answers is correct, given the correctly rearranged code and function calls provided?


In [ ]:
## code to rearrange 
if len(s) == 0:
return s
return r(s[1:]) + s[0]
def r(s)



a. 'gnimmargorp' for r('programming')
b. 'Error for r('programming')
c. 'gnimmargorP' for r('programming')
d. 'prgram' for r(r('programming')[4:])

# answer

In [ ]:
def r(s):
    if len(s) == 0:
        return s
    return r(s[1:]) + s[0]

Let's rearrange the code to create a correct recursive function for reversing a string:



```python
def r(s):
    if len(s) == 0:
        return s
    return r(s[1:]) + s[0]

```

Now, let's analyze the given options to determine which one is correct:

1. `'gnimmargorp' for r('programming')`
   This is correct. The function will reverse 'programming' to 'gnimmargorp'.

2. `'Error for r('programming')`
   This is incorrect. The function will work without errors.

3. `'gnimmargorP' for r('programming')`
   This is incorrect. The 'P' should be at the end, not the beginning.

4. `'prgram' for r(r('programming')[4:])`
   Let's break this down:
   - `r('programming')` gives 'gnimmargorp'
   - `r('programming')[4:]` gives 'margorp'
   - `r('margorp')` gives 'program'

Therefore, the correct answer is:
`'gnimmargorp' for r('programming')`

This recursive function works as follows:
1. If the string is empty, return it (base case).
2. Otherwise, recursively call the function with all but the first character, then append the first character at the end.
3. This process continues until the base case is reached, effectively reversing the string.

Would you like me to explain any part of this in more detail?

# 5.  What is the lambda expression required to filter out negative numbers from a list:


numbers = [-2, 5, -9, 8, -1, 10]
result = filter(, numbers)
print(list(result))

# Output objectives
lambda x: -x
lambda x: x < 0
lambda x: x >= 0
lambda x: x == 0

To filter out negative numbers from a list, we need a lambda expression that returns True for non-negative numbers (i.e., numbers greater than or equal to 0) and False for negative numbers. Let's analyze each option:

lambda x: -x
This simply negates each number, it doesn't filter anything.
lambda x: x < 0
This returns True for negative numbers, which is the opposite of what we want.
lambda x: x >= 0
This returns True for non-negative numbers (0 and positive numbers), which is exactly what we need.
lambda x: x == 0
This only returns True for zero, filtering out all other numbers.

The correct lambda expression to filter out negative numbers from a list is:


`lambda x: x >= 0`

This lambda function returns True for numbers that are greater than or equal to 0 (non-negative numbers), and False for negative numbers, effectively filtering them out.

In [3]:
# Full implementation
numbers = [-2, 5, -9, 8, -1, 10]
result = filter(lambda x: x >= 0, numbers)
print(list(result))


[5, 8, 10]


6. Rearrange the following code to correctly implement a Python class for a Circle, including a method to calculate the circumference.

Which of the following options give the correct answer, given that the radius of the circle is 3?

def __init__(self, radius):
self.radius = radius
return 2 * 3.14159 * self.radius
class Circle:
def circumference(self):
SyntaxError for Circle(3).circumference()
None for Circle().circumference(3)
18.849539999999998 for Circle(3).circumference()
TypeError for Circle().circumference(3)

To correctly implement a Python class for a Circle, including a method to calculate the circumference, the code needs to be arranged logically with the correct class structure, constructor (__init__ method), and the method to calculate the circumference.

Rearranged and corrected code:

In [5]:
class Circle:
    # Constructor to initialize the radius of the circle
    def __init__(self, radius):
        self.radius = radius

    # Method to calculate the circumference
    def circumference(self):
        return 2 * 3.14159 * self.radius


In [6]:
#Example code
# Create an instance of Circle with a radius of 3
circle = Circle(3)

# Call the circumference method
print(circle.circumference())


18.849539999999998


`Explanation of options:`
18.849539999999998 for Circle(3).circumference(): This is the correct output because, for a circle with radius 3, the circumference is calculated as:

Circumference
=
2
×
𝜋
×
radius
=
2
×
3.14159
×
3
=
18.84954
Circumference=2×π×radius=2×3.14159×3=18.84954
SyntaxError for Circle(3).circumference(): There is no syntax error here because the class is properly defined.

None for Circle().circumference(3) and TypeError for Circle().circumference(3): These are incorrect because Circle() requires a radius to be passed during instantiation, and the method does not take any arguments besides self.

7. Rearrange the following code to correctly define a Python class representing a rectangle, which includes methods for calculating area and perimeter. 

What will be the output of the following code?

return 2 * (self.width + self.height)
def perimeter(self):
return self.width * self.height
class Rectangle:
def __init__(self, width, height):
self.width = width
self.height = height
def area(self):

rect = Rectangle(3, 4)
print("Area:", rect.area(), "Perimeter:", rect.perimeter())
Area: 14 Perimeter: 18
Area: 12 Perimeter: 14
TypeError
Area: 12 Perimeter: None

`To correctly define a Python class representing a rectangle with methods to calculate the area and perimeter, the code needs to be rearranged logically to include:`

The constructor (__init__) to initialize the rectangle's width and height.
Methods to calculate the area (area()) and the perimeter (perimeter()).
`Correct implementation:`

In [7]:
class Rectangle:
    # Constructor to initialize width and height
    def __init__(self, width, height):
        self.width = width
        self.height = height

    # Method to calculate the area
    def area(self):
        return self.width * self.height

    # Method to calculate the perimeter
    def perimeter(self):
        return 2 * (self.width + self.height)



In [8]:
# Create an instance of Rectangle with width=3 and height=4
rect = Rectangle(3, 4)

# Print area and perimeter
print("Area:", rect.area(), "Perimeter:", rect.perimeter())


Area: 12 Perimeter: 14


`Explanation:`
The area() method returns the product of width and height: 
3
×
4
=
12
3×4=12.
The perimeter() method returns the sum of twice the width and twice the height: 
2
×
(
3
+
4
)
=
14
2×(3+4)=14.
Explanation of options:
Area: 12 Perimeter: 14: This is the correct output for a rectangle with width 3 and height 4, as shown above.

Area: 14 Perimeter: 18: This is incorrect because the area is computed as width 
×
× height, which is 12, not 14. Similarly, the perimeter is 14, not 18.

TypeError: This would occur if the methods were incorrectly defined or missing arguments, but in this case, the class is correctly defined.

Area: 12 Perimeter: None: This would happen if the perimeter() method did not return a value, but it correctly returns the perimeter.

Final Answer:
The correct output is Area: 12 Perimeter: 14.

8. Rearrange the following lines of code to create a lambda function that squares each number in a list, using the map function:

print(list(result))
result = map(lambda x: x**2, numbers)
numbers = [1, 2, 3, 4, 5]
SyntaxError
[1, 2, 3, 4, 5]
[2, 4, 6, 8, 10]
[1, 4, 9, 16, 25]

To create a lambda function that squares each number in a list using the map function, the code needs to be rearranged in the correct sequence.

Correct rearranged code:

In [9]:
numbers = [1, 2, 3, 4, 5]
result = map(lambda x: x**2, numbers)
print(list(result))


[1, 4, 9, 16, 25]


`Explanation:`
numbers = [1, 2, 3, 4, 5]: This defines the list of numbers.
map(lambda x: x**2, numbers): This applies the lambda function (which squares each element) to each number in the list.
print(list(result)): This converts the result (which is an iterator) to a list and prints it.

9. Identify the line that violates the PEP 8 style guidelines:

class ExampleClass:
def __init__(self,value):
self.value = value
def get_value(self):
return self.value
def __init__(self,value):
class ExampleClass:
return self.value
self.value = value
def get_value(self):

`The line that violates the PEP 8 style guidelines is:`

In [ ]:
def __init__(self,value): 
    

Violation:
Missing space after the comma between the parameters self and value. According to PEP 8, function arguments should be separated by a single space after each comma.
Corrected version:

In [ ]:
def __init__(self, value):


`Full corrected code with proper PEP 8 formatting:`

In [ ]:
class ExampleClass:
    def __init__(self, value):
        self.value = value

    def get_value(self):
        return self.value

`Explanation:`
Indentation: Each function inside the class should be indented by four spaces.
Spaces around arguments: PEP 8 recommends using a space after each comma in argument lists for readability.

10. The following Python code is meant to calculate the total area of two different shapes, but it contains an error. Identify and correct the error.


class Shape:
    def __init__(self, name):
        self.name = name

class Circle(Shape):
    def __init__(self, radius):
        super().__init__("Circle")
        self.radius = radius

    def area(self):
        return 3.14 * self.radius * self.radius
class Square(Shape):
    def __init__(self, side):
        super().__init__("Square")
        self.side = side

    def area(self):
        return self.side * self.side

circle = Circle(3)
square = Square(4)
total_area = circle.area + square.area
print("Total Area:", total_area)
 

a. There is no error in the code.
b. The error is a mistake in the Circle class's _init_ method. The super().__init__("Circle") call should come after initialising the radius attribute.
c. The error is a missing self parameter in the area method of the Square class.
d. The error is that circle.area and square.area are method references, not method calls.

`The error in the code is:`

`"The error is that circle.area and square.area are method references, not method calls."`

Explanation:
In Python, circle.area and square.area without parentheses refer to the method itself, not the result of calling the method. To call the method and get the area, you need to use parentheses: circle.area() and square.area().

In [10]:
class Shape:
    def __init__(self, name):
        self.name = name

class Circle(Shape):
    def __init__(self, radius):
        super().__init__("Circle")
        self.radius = radius

    def area(self):
        return 3.14 * self.radius * self.radius

class Square(Shape):
    def __init__(self, side):
        super().__init__("Square")
        self.side = side

    def area(self):
        return self.side * self.side

circle = Circle(3)
square = Square(4)
total_area = circle.area() + square.area()  # Corrected: method calls
print("Total Area:", total_area)


Total Area: 44.26


Now, let's examine each of the provided options:

There is no error in the code.
`This is incorrect. There is an error in the code.`
The error is a mistake in the Circle class's init method. The super().init("Circle") call should come after initialising the radius attribute.
`This is incorrect. The order of initialization in the Circle class is fine.`
The error is a missing self parameter in the area method of the Square class.
`This is incorrect. Both area methods are correctly defined with the self parameter.`
The error is that circle.area and square.area are method references, not method calls.
`This is correct. This is the error in the code.`